### PR TITLE
Local authentication plugin

### PR DESCRIPTION
**Pull request by mehmetf**
This is the first plugin with its own resources to be checked in AFAIK.

Please review and see whether you are comfortable with i18n'ed strings being stored this way. Our current proposal is for the app using this to embed strings into their own .arb bundles and get them translated. Unfortunately, we know of no way of shipping translated strings with a plugin since merging these at runtime seems to be infeasible.*

*: One way is to provide the strings at native layer and translate them. However, that creates a large burden on plugin maintainer [need to translate both iOS and Android flavors] and also makes overriding the string value more complicated.

### CODE REVIEW

**Comment by tvolkert** [comment]
I'd like @alan-knight to have a look and chime in.

Alan, check out [`auth_strings.dart`](https://github.com/flutter/plugins/pull/151/files#diff-206b7ea9348f3767a0f1033a912b68f3) and how it's used in [`local_auth.dart`](https://github.com/flutter/plugins/pull/151/files#diff-f997d8930094f3111f91642a1d40d5d8)

**Comment by eseidelGoogle** [comment]
How do Cocoapods or Gradle plugins provide translations today?  e.g. does the FaceBook SDK come with pre-translated strings, or is it up to every app developer to send off all their own strings + all the strings for their various dependencies off for translation?  @cbracken, @collinjackson or @jason-simmons might know.

I suspect that we will need to change the way package:intl works for Flutter to use the default iOS and Android data file formats, but I don't know?

**Comment by alan-knight** [comment]
I'm suspicious that putting the messages in a Map like that at initialization could run into a race condition between locale initialization and Map creation. That is, if that Map gets created before initializeMessages has completed, then they will have the wrong language. And obviously if the user changes the language then they'll be wrong.

I guess the obvious alternatives would be to do it as a case statement form, to wrap each method in a switch between the iOS and Android versions. Or possibly you could make the map be a map of closures. So e.g. instead of a getter, make the messages be methods/functions.

      fingerprintHint() => Intl.message ... 

then you could have a map 

     'fingerprintHint' : fingerprintHint,
      ...

and you just have to invoke the closure once you get it out of the map. And be sure that that only happens after locale initialization, but you'd have more control there.

I have a half-formed thought about being able to have the plugin have its own separate messages, involving tweaking the locale name and zones (or maybe just wrapping functions).

**Comment by mehmetf** [comment]
@alan-knight Thanks for the comments.

We are not pulling the strings during initialization of the plugin or the app. It is done when `.args` getter is called. See https://github.com/flutter/plugins/pull/151/files#diff-f997d8930094f3111f91642a1d40d5d8R47. There should not be a race condition here and even when language changes, the next call to `authenticateWithBiometrics` should evalute all the strings again.

File: packages/local_auth/pubspec.yaml [hunk 1]
```diff
@@ -0,0 +1,18 @@
+name: local_auth
+description: Local authentication plugin for Flutter.
```
Comment (goderbauer) [inline packages/local_auth/pubspec.yaml]: Can this describe better what local authentication is? Maybe mention fingerprint, biometric, etc.

If you scan the list on https://pub.dartlang.org/flutter/packages and read the current description you'll have no idea how cool this plugin actually is. 😄

File: packages/local_auth/lib/local_auth.dart [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+/// A Flutter plugin for authenticating the user identity locally.
+class LocalAuthentication {
+  /// Authenticates the user with biometrics available on the device.
+  ///
+  /// Returns a [Future] holding true, if the user successfully authenticated,
+  /// false otherwise.
+  ///
+  /// Pass useErrorDialogs = false if you don't want to use default error
+  /// dialogs. In that case, errors that can be handled by plugin will be
```
Comment (goderbauer) [inline packages/local_auth/lib/local_auth.dart]: Would be cool to document: 
* What kind of errors would I have to handle manually if set to false?
* What get's automatically handled if set to true?

File: packages/local_auth/lib/local_auth.dart [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+  ///
+  /// Construct [AndroidAuthStrings] and [IOSAuthStrings] if you want to
+  /// customize messages in the dialogs.
+  ///
+  /// Throws an [PlatformException] if there were technical problems with local
+  /// authentication (e.g. lack of relevant hardware). This might throw
+  /// [PlatformException] with error code [otherOperatingSystem] on the iOS
+  /// simulator.
+  Future<bool> authenticateWithBiometrics({
+    @required String localizedReason,
```
Comment (goderbauer) [inline packages/local_auth/lib/local_auth.dart]: The method's doc should probably mention what localizedReason is (since its required).

File: packages/local_auth/pubspec.yaml (hunk 1 above)
Comment (mehmetf) [inline packages/local_auth/pubspec.yaml]: Yep will do. The naming of this plugin is actually very tricky. On iOS local authentication is: https://developer.apple.com/documentation/localauthentication. However on Android such a thing is not possible. You have to do everything manually. For instance we have no PIN support. Hopefully android will improve its API and we can get something closer to iOS.

File: packages/local_auth/lib/local_auth.dart (hunk 3 above)
Comment (mehmetf) [inline packages/local_auth/lib/local_auth.dart]: Done

**Comment by mehmetf** [comment]
Friendly ping!

Are there any other AIs for this PR?

**Comment by tvolkert** [comment]
My gut is that plugins should really be bundling their own translated messages.  As an example, when an app uses the GoogleSignIn plugin, the app developer doesn't expect to provide translations for all the native sign-in UI dialogs and such.

There's a logistical question about how we might go about wiring this into Google's internal process for getting strings translated, but in the meantime, I think it's probably best to structure the API to assume that the translations are self-contained in this plugin (even if it means we'll only have `en` in the first commit).

WDYT?

**Comment by collinjackson** [comment]
It's easier to add to the Dart API than subtract; if we can get away with making the strings self-contained in the plugin, that makes sense to me as a first step.

We could provide an optional mechanism for overriding the string values, either now or in the future, which is what other SDKs like Facebook do.

**Comment by alan-knight** [comment]
Ah, sorry, I didn't understand what was going on. It seems a bit hard to follow, and potentially slow. So we're eagerly building up a map of of all possible messages, keyed by message name, and passing that in to the constructor, which will then look them up by string again? And we have subclasses for Android and for iOS, why not just use the dispatch there to distinguish them?

I have an example of having a component with its own separate, bundled messages. It's not entirely pretty, but it works. What happens is that the component defines its own locale suffix, and then things within the component use that locale. Either they run inside an Intl.withLocale() call, or they explicitly pass in the locale. Not externally visible yet, but shortly, or ping me for the internal reference.

**Comment by mehmetf** [comment]
I agree that providing a complete package makes the most sense. The question is should it be done on Dart side or native side.

Dart side pros:
- A single language/place to define all the strings. More dart, less native code. Hurray!
- Easier for user to override.
- One-time translation and single process.
- Dart translation is well understood by most contributors (who need to translate their own apps).
- Presumably easier to roll into Google.

Dart side cons:
- No proper support for having multiple translation bundles (@alan-knight is working on a solution)
- Not entirely clear where the strings are used. Need to completely rely on string descriptions.
- More keys to maintain.

I am leaning towards Dart side even though I don't think it is a common use case to override these strings. What do the rest of you think?

**Comment by eseidelGoogle** [comment]
Many plugin translations will be outside of our control.  For example, if you have a Example.com SDK plugin, that will likely wrap existing Example.com iOS and Android "native" SDKs and include translations in those formats.  So it is my belief that Flutter (and thus package:intl?) will need to support reading/loading translations from at-least "native" plugin components, but also likely from in-app "native" translation systems.

**Comment by alan-knight** [comment]
It should not be particularly difficult to make it look up messages in some other format. The current generated code imports message_lookup_by_library.dart, which has a reasonably narrow set of messages, just provide a different implementation that knows how to read and interpret the native messages. But there are probably interesting edge cases, like how the messages are identified or if the semantics of the native messages aren't a close match to Dart/ICU semantics.

File: packages/local_auth/lib/local_auth.dart [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    bool useErrorDialogs: true,
+    AndroidAuthMessages androidAuthStrings: const AndroidAuthMessages(),
```
Comment (mehmetf) [inline packages/local_auth/lib/local_auth.dart]: @alan-knight  `const AndroidAuthMessages` does not initialize any map (so we are not eagerly building anything). The app can choose to override a subset of messages by passing in an instance of `AndroidAuthMessages`. When `.args` gets accessed below, the map is built by using a combination of user overrides (if any) and defaults. The defaults come from getters which point to Intl'ed strings at which point the Intl package evaluates the locale and sets the correct default.

Perhaps I am missing an optimization here. Please let me know if there's a better way.

**Comment by tvolkert** [comment]
@alan-knight it sounds like the workaround of the component defining its own locale suffix is a temporary workaround.  As we get more and more plugins (and third-party widgets with built-in strings), this will become a pattern that's seen again and again.

As such, can we add first-class support of multiple message bundles to `package:intl`?  It looks like it's covered by https://github.com/dart-lang/i18n/issues/358.

**Comment by mehmetf** [comment]
@eseidelGoogle I don't foresee issues with bundling of native strings. If you have pre-translated bundles, it should just work due to gradle manifest/res merge on Android side. I am not sure about iOS mechanism but I have evidence that works also (internal Feedback plugin).

**Comment by tvolkert** [comment]
So at this point, my understanding is that our choices are:

* Put the messages in Dart (our desired solution), but in so doing, we need to employ the workaround of having the plugin define its own locate suffix (and we require a new `package:intl` release that enables the workaround)
* Put the messages in native bundles, which unfortunately means that we'll have duplicate translated strings (for iOS and Android).  This in turn means:
  * The two sets of translations could get out of sync
  * We'll have the extra overhead of creating and maintaining the duplicate bundles (the process for each platform is different), which will be burdensome on plugin authors.

**Comment by mehmetf** [comment]
Since we expect the multi-bundle problem to be resolved in Dart, can we make a bet on it and let the PR go in as is? Until Dart has proper support for this, the recommended workaround is for apps to manually merge the strings in the plugin to their bundle. @tvolkert @collinjackson are you OK with this?

@alan-knight What would be the timeline to resolve dart-lang/i18n#358?

**Comment by alan-knight** [comment]
@tvolkert You don't need a new Intl release to do the trick with a locale suffix. The unpublished part is just an example/test for how to do it, there are no code changes to Intl.

@mehmetf Right now there is no timeline for resolving that. It's potentially a significant change. We might be able to do something in code generation with exposing a different Intl class, but it's certainly not worked out.

**Comment by tvolkert** [comment]
LGTM on the initial commit.  I think we should follow-up with a change to use the locale suffix trick suggested by @alan-knight before we publish the plugin.
